File: lite/examples/model_personalization/android/app/src/main/java/org/tensorflow/lite/examples/transfer/CameraFragment.java
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

package org.tensorflow.lite.examples.transfer;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Bitmap.Config;
import android.graphics.Canvas;
import android.graphics.ImageFormat;
import android.graphics.Matrix;
import android.os.Bundle;
import android.os.Handler;
import android.os.HandlerThread;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.Rational;
import android.util.Size;
import android.view.Display;
import android.view.LayoutInflater;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.camera.core.AspectRatio;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.CameraX;
import androidx.camera.core.ImageAnalysis;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageProxy;
import androidx.camera.core.ImageProxy.PlaneProxy;
import androidx.camera.core.Preview;
import androidx.camera.core.impl.ImageAnalysisConfig;
import androidx.camera.core.impl.PreviewConfig;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.content.ContextCompat;
import androidx.databinding.BindingAdapter;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.ViewModelProviders;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.android.material.snackbar.Snackbar;
import com.google.common.util.concurrent.ListenableFuture;

import java.nio.ByteBuffer;
import java.util.Locale;
import java.util.Objects;
import java.util.UUID;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import org.tensorflow.lite.examples.transfer.api.TransferLearningModel.Prediction;
import org.tensorflow.lite.examples.transfer.databinding.CameraFragmentBinding;

/**
 * The main fragment of the classifier.
 *
 * Camera functionality (through CameraX) is heavily based on the official example:
 * https://github.com/android/camera/tree/master/CameraXBasic.
 */
public class CameraFragment extends Fragment {

  private static final int LOWER_BYTE_MASK = 0xFF;

  private static final String TAG = CameraFragment.class.getSimpleName();

  private PreviewView viewFinder;

  private Integer viewFinderRotation = null;
  private int lensFacing = CameraSelector.LENS_FACING_BACK;

  private Size bufferDimens = new Size(0, 0);
  private Size viewFinderDimens = new Size(0, 0);

  private CameraFragmentViewModel viewModel;
  private TransferLearningModelWrapper tlModel;

  // When the user presses the "add sample" button for some class,
  // that class will be added to this queue. It is later extracted by
  // InferenceThread and processed.
  private final ConcurrentLinkedQueue<String> addSampleRequests = new ConcurrentLinkedQueue<>();

  private final LoggingBenchmark inferenceBenchmark = new LoggingBenchmark("InferenceBench");

  private ExecutorService cameraExecutor = Executors.newFixedThreadPool(1);

  private Integer rotation;

  private final double RATIO_4_3_VALUE = 4.0 / 3.0;
  private final double RATIO_16_9_VALUE = 16.0 / 9.0;

  private int aspectRatio(int width, int height) {
    final double previewRatio = (double)Math.max(width, height) / (double)Math.min(width, height);
    if (Math.abs(previewRatio - RATIO_4_3_VALUE) <= Math.abs(previewRatio - RATIO_16_9_VALUE)) {
      return AspectRatio.RATIO_4_3;
    }
    return AspectRatio.RATIO_16_9;
  }

  /**
   * Set up a responsive preview for the view finder.
   */
  private void startCamera() {
    viewFinderRotation = getDisplaySurfaceRotation(viewFinder.getDisplay());
    if (viewFinderRotation == null) {
      viewFinderRotation = 0;
    }

    DisplayMetrics metrics = new DisplayMetrics();
    viewFinder.getDisplay().getRealMetrics(metrics);

    int screenAspectRatio = aspectRatio(metrics.widthPixels, metrics.heightPixels);
    rotation = getDisplaySurfaceRotation(viewFinder.getDisplay());

    CameraSelector cameraSelector = new CameraSelector.Builder()
            .requireLensFacing(lensFacing).build();
    ListenableFuture<ProcessCameraProvider> cameraProviderFuture =
            ProcessCameraProvider.getInstance(requireContext());

    cameraProviderFuture.addListener(()->{

      // CameraProvider
      ProcessCameraProvider cameraProvider = null;
      try {
        cameraProvider = cameraProviderFuture.get();
      } catch (ExecutionException e) {
        e.printStackTrace();
      } catch (InterruptedException e) {
        e.printStackTrace();
      }

      // Preview
      Preview preview = new Preview.Builder()
              // We request aspect ratio but no resolution
              .setTargetAspectRatio(screenAspectRatio)
              // Set initial target rotation
              .setTargetRotation(rotation)
              .build();

      // Attach the viewfinder's surface provider to preview use case
      preview.setSurfaceProvider(viewFinder.getPreviewSurfaceProvider());

      ImageAnalysis imageAnalyzer = new ImageAnalysis.Builder()
              .setTargetAspectRatio(screenAspectRatio)
              .setTargetRotation(rotation)
              .build();
      imageAnalyzer.setAnalyzer(cameraExecutor, inferenceAnalyzer);

      // Must unbind the use-cases before rebinding them
      cameraProvider.unbindAll();

      try {
        // A variable number of use-cases can be passed here -
        // camera provides access to CameraControl & CameraInfo
        Camera camera = cameraProvider.bindToLifecycle(
                this, cameraSelector, preview, imageAnalyzer);
      } catch(Exception e) {
        Log.e(TAG, "Use case binding failed", e);
      }

    }, ContextCompat.getMainExecutor(requireContext()));
  }

  private final ImageAnalysis.Analyzer inferenceAnalyzer =
      (imageProxy) -> {
        final String imageId = UUID.randomUUID().toString();

        inferenceBenchmark.startStage(imageId, "preprocess");
        float[] rgbImage = prepareCameraImage(yuvCameraImageToBitmap(imageProxy), rotation);
        inferenceBenchmark.endStage(imageId, "preprocess");

        // Adding samples is also handled by inference thread / use case.
        // We don't use CameraX ImageCapture since it has very high latency (~650ms on Pixel 2 XL)
        // even when using .MIN_LATENCY.
        String sampleClass = addSampleRequests.poll();
        if (sampleClass != null) {
          inferenceBenchmark.startStage(imageId, "addSample");
          try {
            tlModel.addSample(rgbImage, sampleClass).get();
          } catch (ExecutionException e) {
            throw new RuntimeException("Failed to add sample to model", e.getCause());
          } catch (InterruptedException e) {
            // no-op
          }

          viewModel.increaseNumSamples(sampleClass);
          inferenceBenchmark.endStage(imageId, "addSample");

        } else {
          // We don't perform inference when adding samples, since we should be in capture mode
          // at the time, so the inference results are not actually displayed.
          inferenceBenchmark.startStage(imageId, "predict");
          Prediction[] predictions = tlModel.predict(rgbImage);
          if (predictions == null) {
            return;
          }
          inferenceBenchmark.endStage(imageId, "predict");

          for (Prediction prediction : predictions) {
            viewModel.setConfidence(prediction.getClassName(), prediction.getConfidence());
          }
        }

        imageProxy.close();
        inferenceBenchmark.finish(imageId);
      };

  public final View.OnClickListener onAddSampleClickListener = view -> {
    String className;
    if (view.getId() == R.id.class_btn_1) {
      className = "1";
    } else if (view.getId() == R.id.class_btn_2) {
      className = "2";
    } else if (view.getId() == R.id.class_btn_3) {
      className = "3";
    } else if (view.getId() == R.id.class_btn_4) {
      className = "4";
    } else {
      throw new RuntimeException("Listener called for unexpected view");
    }

    addSampleRequests.add(className);
  };

  @Override
  public void onCreate(Bundle bundle) {
    super.onCreate(bundle);

    tlModel = new TransferLearningModelWrapper(getActivity());
    viewModel = ViewModelProviders.of(this).get(CameraFragmentViewModel.class);
    viewModel.setTrainBatchSize(tlModel.getTrainBatchSize());
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle bundle) {
    CameraFragmentBinding dataBinding =
        DataBindingUtil.inflate(inflater, R.layout.camera_fragment, container, false);
    dataBinding.setLifecycleOwner(getViewLifecycleOwner());
    dataBinding.setVm(viewModel);
    View rootView = dataBinding.getRoot();

    for (int buttonId : new int[] {
        R.id.class_btn_1, R.id.class_btn_2, R.id.class_btn_3, R.id.class_btn_4}) {
      rootView.findViewById(buttonId).setOnClickListener(onAddSampleClickListener);
    }

    ChipGroup chipGroup = (ChipGroup) rootView.findViewById(R.id.mode_chip_group);
    if (viewModel.getCaptureMode().getValue()) {
      ((Chip) rootView.findViewById(R.id.capture_mode_chip)).setChecked(true);
    } else {
      ((Chip) rootView.findViewById(R.id.inference_mode_chip)).setChecked(true);
    }

    chipGroup.setOnCheckedChangeListener((group, checkedId) -> {
      if (checkedId == R.id.capture_mode_chip) {
        viewModel.setCaptureMode(true);
      } else if (checkedId == R.id.inference_mode_chip) {
        viewModel.setCaptureMode(false);
      }
    });

    return dataBinding.getRoot();
  }

  @Override
  public void onViewCreated(View view, Bundle bundle) {
    super.onViewCreated(view, bundle);

    viewFinder = getActivity().findViewById(R.id.view_finder);
    viewFinder.post(this::startCamera);
  }

  @Override
  public void onActivityCreated(Bundle bundle) {
    super.onActivityCreated(bundle);

    viewModel
        .getTrainingState()
        .observe(
            getViewLifecycleOwner(),
            trainingState -> {
              switch (trainingState) {
                case STARTED:
                  tlModel.enableTraining((epoch, loss) -> viewModel.setLastLoss(loss));
                  if (!viewModel.getInferenceSnackbarWasDisplayed().getValue()) {
                    Snackbar.make(
                            getActivity().findViewById(R.id.classes_bar),
                            R.string.switch_to_inference_hint,
                            Snackbar.LENGTH_LONG)
                        .show();
                    viewModel.markInferenceSnackbarWasCalled();
                  }
                  break;
                case PAUSED:
                  tlModel.disableTraining();
                  break;
                case NOT_STARTED:
                  break;
              }
            });
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    tlModel.close();
    tlModel = null;
  }

  private static Integer getDisplaySurfaceRotation(Display display) {
    if (display == null) {
      return null;
    }

    switch (display.getRotation()) {
      case Surface.ROTATION_0: return 0;
      case Surface.ROTATION_90: return 90;
      case Surface.ROTATION_180: return 180;
      case Surface.ROTATION_270: return 270;
      default: return null;
    }
  }

  private static Bitmap yuvCameraImageToBitmap(ImageProxy imageProxy) {
    if (imageProxy.getFormat() != ImageFormat.YUV_420_888) {
      throw new IllegalArgumentException(
          "Expected a YUV420 image, but got " + imageProxy.getFormat());
    }

    PlaneProxy yPlane = imageProxy.getPlanes()[0];
    PlaneProxy uPlane = imageProxy.getPlanes()[1];

    int width = imageProxy.getWidth();
    int height = imageProxy.getHeight();

    byte[][] yuvBytes = new byte[3][];
    int[] argbArray = new int[width * height];
    for (int i = 0; i < imageProxy.getPlanes().length; i++) {
      final ByteBuffer buffer = imageProxy.getPlanes()[i].getBuffer();
      yuvBytes[i] = new byte[buffer.capacity()];
      buffer.get(yuvBytes[i]);
    }

    ImageUtils.convertYUV420ToARGB8888(
        yuvBytes[0],
        yuvBytes[1],
        yuvBytes[2],
        width,
        height,
        yPlane.getRowStride(),
        uPlane.getRowStride(),
        uPlane.getPixelStride(),
        argbArray);

    return Bitmap.createBitmap(argbArray, width, height, Config.ARGB_8888);
  }

  /**
   * Normalizes a camera image to [0; 1], cropping it
   * to size expected by the model and adjusting for camera rotation.
   */
  private static float[] prepareCameraImage(Bitmap bitmap, int rotationDegrees)  {
    int modelImageSize = TransferLearningModelWrapper.IMAGE_SIZE;

    Bitmap paddedBitmap = padToSquare(bitmap);
    Bitmap scaledBitmap = Bitmap.createScaledBitmap(
        paddedBitmap, modelImageSize, modelImageSize, true);

    Matrix rotationMatrix = new Matrix();
    rotationMatrix.postRotate(rotationDegrees);
    Bitmap rotatedBitmap = Bitmap.createBitmap(
        scaledBitmap, 0, 0, modelImageSize, modelImageSize, rotationMatrix, false);

    float[] normalizedRgb = new float[modelImageSize * modelImageSize * 3];
    int nextIdx = 0;
    for (int y = 0; y < modelImageSize; y++) {
      for (int x = 0; x < modelImageSize; x++) {
        int rgb = rotatedBitmap.getPixel(x, y);

        float r = ((rgb >> 16) & LOWER_BYTE_MASK) * (1 / 255.f);
        float g = ((rgb >> 8) & LOWER_BYTE_MASK) * (1 / 255.f);
        float b = (rgb & LOWER_BYTE_MASK) * (1 / 255.f);

        normalizedRgb[nextIdx++] = r;
        normalizedRgb[nextIdx++] = g;
        normalizedRgb[nextIdx++] = b;
      }
    }

    return normalizedRgb;
  }

  private static Bitmap padToSquare(Bitmap source) {
    int width = source.getWidth();
    int height = source.getHeight();

    int paddingX = width < height ? (height - width) / 2 : 0;
    int paddingY = height < width ? (width - height) / 2 : 0;
    Bitmap paddedBitmap = Bitmap.createBitmap(
        width + 2 * paddingX, height + 2 * paddingY, Config.ARGB_8888);
    Canvas canvas = new Canvas(paddedBitmap);
    canvas.drawARGB(0xFF, 0xFF, 0xFF, 0xFF);
    canvas.drawBitmap(source, paddingX, paddingY, null);
    return paddedBitmap;
  }

  // Binding adapters:

  @BindingAdapter({"captureMode", "inferenceText", "captureText"})
  public static void setClassSubtitleText(
      TextView view, boolean captureMode, Float inferenceText, Integer captureText) {
    if (captureMode) {
      view.setText(captureText != null ? Integer.toString(captureText) : "0");
    } else {
      view.setText(
          String.format(Locale.getDefault(), "%.2f", inferenceText != null ? inferenceText : 0.f));
    }
  }

  @BindingAdapter({"android:visibility"})
  public static void setViewVisibility(View view, boolean visible) {
    view.setVisibility(visible ? View.VISIBLE : View.GONE);
  }

  @BindingAdapter({"highlight"})
  public static void setClassButtonHighlight(View view, boolean highlight) {
    int drawableId;
    if (highlight) {
      drawableId = R.drawable.btn_default_highlight;
    } else {
      drawableId = R.drawable.btn_default;
    }
    view.setBackground(view.getContext().getDrawable(drawableId));
  }
}

File: lite/examples/model_personalization/android/app/src/main/java/org/tensorflow/lite/examples/modelpersonalization/fragments/CameraFragment.kt
/*
 * Copyright 2022 The TensorFlow Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *             http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tensorflow.lite.examples.modelpersonalization.fragments

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.graphics.Bitmap
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.SystemClock
import android.util.Log
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.content.res.AppCompatResources
import androidx.camera.core.Preview
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.Camera
import androidx.camera.core.CameraSelector
import androidx.camera.core.AspectRatio
import androidx.camera.core.ImageProxy
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.Navigation
import org.tensorflow.lite.examples.modelpersonalization.MainViewModel
import org.tensorflow.lite.examples.modelpersonalization.R
import org.tensorflow.lite.examples.modelpersonalization.TransferLearningHelper
import org.tensorflow.lite.examples.modelpersonalization.TransferLearningHelper.Companion.CLASS_FOUR
import org.tensorflow.lite.examples.modelpersonalization.TransferLearningHelper.Companion.CLASS_ONE
import org.tensorflow.lite.examples.modelpersonalization.TransferLearningHelper.Companion.CLASS_THREE
import org.tensorflow.lite.examples.modelpersonalization.TransferLearningHelper.Companion.CLASS_TWO
import org.tensorflow.lite.examples.modelpersonalization.databinding.FragmentCameraBinding
import org.tensorflow.lite.support.label.Category
import java.util.Locale
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class CameraFragment : Fragment(),
    TransferLearningHelper.ClassifierListener {

    companion object {
        private const val TAG = "Model Personalization"
        private const val LONG_PRESS_DURATION = 500
        private const val SAMPLE_COLLECTION_DELAY = 300
    }

    private var _fragmentCameraBinding: FragmentCameraBinding? = null
    private val fragmentCameraBinding
        get() = _fragmentCameraBinding!!

    private lateinit var transferLearningHelper: TransferLearningHelper
    private lateinit var bitmapBuffer: Bitmap

    private val viewModel: MainViewModel by activityViewModels()
    private var preview: Preview? = null
    private var imageAnalyzer: ImageAnalysis? = null
    private var camera: Camera? = null
    private var cameraProvider: ProcessCameraProvider? = null
    private var previousClass: String? = null

    /** Blocking camera operations are performed using this executor */
    private lateinit var cameraExecutor: ExecutorService

    // When the user presses the "add sample" button for some class,
    // that class will be added to this queue. It is later extracted by
    // InferenceThread and processed.
    private val addSampleRequests = ConcurrentLinkedQueue<String>()

    private var sampleCollectionButtonPressedTime: Long = 0
    private var isCollectingSamples = false
    private val sampleCollectionHandler = Handler(Looper.getMainLooper())
    private val onAddSampleTouchListener =
        View.OnTouchListener { view, motionEvent ->
            when (motionEvent.action) {
                MotionEvent.ACTION_DOWN -> {
                    isCollectingSamples = true
                    sampleCollectionButtonPressedTime =
                        SystemClock.uptimeMillis()
                    sampleCollectionHandler.post(object : Runnable {
                        override fun run() {
                            val timePressed =
                                SystemClock.uptimeMillis() - sampleCollectionButtonPressedTime
                            view.findViewById<View>(view.id).performClick()
                            if (timePressed < LONG_PRESS_DURATION) {
                                sampleCollectionHandler.postDelayed(
                                    this,
                                    LONG_PRESS_DURATION.toLong()
                                )
                            } else if (isCollectingSamples) {
                                val className: String =
                                    getClassNameFromResourceId(view.id)
                                addSampleRequests.add(className)
                                sampleCollectionHandler.postDelayed(
                                    this,
                                    SAMPLE_COLLECTION_DELAY.toLong()
                                )
                            }
                        }
                    })
                }
                MotionEvent.ACTION_UP -> {
                    sampleCollectionHandler.removeCallbacksAndMessages(null)
                    isCollectingSamples = false
                }
            }
            true
        }

    override fun onResume() {
        super.onResume()

        if (!PermissionsFragment.hasPermissions(requireContext())) {
            Navigation.findNavController(
                requireActivity(),
                R.id.fragment_container
            ).navigate(CameraFragmentDirections.actionCameraToPermissions())
        }
    }

    override fun onDestroyView() {
        _fragmentCameraBinding = null
        super.onDestroyView()

        // Shut down our background executor
        cameraExecutor.shutdown()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _fragmentCameraBinding =
            FragmentCameraBinding.inflate(inflater, container, false)

        return fragmentCameraBinding.root
    }

    @SuppressLint("MissingPermission", "ClickableViewAccessibility")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        transferLearningHelper = TransferLearningHelper(
            context = requireContext(),
            classifierListener = this
        )

        cameraExecutor = Executors.newSingleThreadExecutor()

        viewModel.numThreads.observe(viewLifecycleOwner) {
            transferLearningHelper.numThreads = it
            transferLearningHelper.close()
            if (viewModel.getTrainingState() != MainViewModel.TrainingState.PREPARE) {
                // If the model is training, continue training with old image
                // sets.
                viewModel.setTrainingState(MainViewModel.TrainingState.TRAINING)
                transferLearningHelper.startTraining()
            }
        }

        viewModel.trainingState.observe(viewLifecycleOwner) {
            updateTrainingButtonState()
        }

        viewModel.captureMode.observe(viewLifecycleOwner) { isCaptureMode ->
            if (isCaptureMode) {
                viewModel.getNumberOfSample()?.let {
                    updateNumberOfSample(it)
                }
                // Unhighlight all class buttons
                highlightResult(null)
            }

            // Update the UI after switch to training mode.
            updateTrainingButtonState()
        }

        viewModel.numberOfSamples.observe(viewLifecycleOwner) {
            // Update the number of samples
            updateNumberOfSample(it)
            updateTrainingButtonState()
        }

        with(fragmentCameraBinding) {
            if (viewModel.getCaptureMode()!!) {
                btnTrainingMode.isChecked = true
            } else {
                btnInferenceMode.isChecked = true
            }
            llClassOne.setOnClickListener {
                addSampleRequests.add(CLASS_ONE)
            }
            llClassTwo.setOnClickListener {
                addSampleRequests.add(CLASS_TWO)
            }
            llClassThree.setOnClickListener {
                addSampleRequests.add(CLASS_THREE)
            }
            llClassFour.setOnClickListener {
                addSampleRequests.add(CLASS_FOUR)
            }
            llClassOne.setOnTouchListener(onAddSampleTouchListener)
            llClassTwo.setOnTouchListener(onAddSampleTouchListener)
            llClassThree.setOnTouchListener(onAddSampleTouchListener)
            llClassFour.setOnTouchListener(onAddSampleTouchListener)
            btnPauseTrain.setOnClickListener {
                viewModel.setTrainingState(MainViewModel.TrainingState.PAUSE)
                transferLearningHelper.pauseTraining()
            }
            btnResumeTrain.setOnClickListener {
                viewModel.setTrainingState(MainViewModel.TrainingState.TRAINING)
                transferLearningHelper.startTraining()
            }
            btnStartTrain.setOnClickListener {
                // Start training process
                viewModel.setTrainingState(MainViewModel.TrainingState.TRAINING)
                transferLearningHelper.startTraining()
            }
            radioButton.setOnCheckedChangeListener { _, checkedId ->
                if (checkedId == R.id.btnTrainingMode) {
                    // Switch to training mode.
                    viewModel.setCaptureMode(true)
                } else {
                    if (viewModel.getTrainingState() == MainViewModel.TrainingState.PREPARE) {
                        fragmentCameraBinding.btnTrainingMode.isChecked = true
                        fragmentCameraBinding.btnInferenceMode.isChecked = false

                        Toast.makeText(
                            requireContext(), "Inference can only " +
                                    "start after training is done.", Toast
                                .LENGTH_LONG
                        ).show()
                    } else {
                        // Pause the training process and switch to inference mode.
                        transferLearningHelper.pauseTraining()
                        viewModel.setTrainingState(MainViewModel.TrainingState.PAUSE)
                        viewModel.setCaptureMode(false)
                    }
                }
            }

            viewFinder.post {
                // Set up the camera and its use cases
                setUpCamera()
            }
        }
    }

    // Initialize CameraX, and prepare to bind the camera use cases
    private fun setUpCamera() {
        val cameraProviderFuture =
            ProcessCameraProvider.getInstance(requireContext())
        cameraProviderFuture.addListener(
            {
                // CameraProvider
                cameraProvider = cameraProviderFuture.get()

                // Build and bind the camera use cases
                bindCameraUseCases()
            },
            ContextCompat.getMainExecutor(requireContext())
        )
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        imageAnalyzer?.targetRotation =
            fragmentCameraBinding.viewFinder.display.rotation
    }

    // Declare and bind preview, capture and analysis use cases
    @SuppressLint("UnsafeOptInUsageError")
    private fun bindCameraUseCases() {

        // CameraProvider
        val cameraProvider =
            cameraProvider
                ?: throw IllegalStateException("Camera initialization failed.")

        // CameraSelector - makes assumption that we're only using the back camera
        val cameraSelector =
            CameraSelector.Builder()
                .requireLensFacing(CameraSelector.LENS_FACING_BACK).build()

        // Preview. Only using the 4:3 ratio because this is the closest to our models
        preview =
            Preview.Builder()
                .setTargetAspectRatio(AspectRatio.RATIO_4_3)
                .setTargetRotation(fragmentCameraBinding.viewFinder.display.rotation)
                .build()

        // ImageAnalysis. Using RGBA 8888 to match how our models work
        imageAnalyzer =
            ImageAnalysis.Builder()
                .setTargetAspectRatio(AspectRatio.RATIO_4_3)
                .setTargetRotation(fragmentCameraBinding.viewFinder.display.rotation)
                .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
                .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
                .build()
                // The analyzer can then be assigned to the instance
                .also {
                    it.setAnalyzer(cameraExecutor) { image ->
                        if (!::bitmapBuffer.isInitialized) {
                            // The image rotation and RGB image buffer are initialized only once
                            // the analyzer has started running
                            bitmapBuffer = Bitmap.createBitmap(
                                image.width,
                                image.height,
                                Bitmap.Config.ARGB_8888
                            )
                        }

                        val sampleClass = addSampleRequests.poll()
                        if (sampleClass != null) {
                            addSample(image, sampleClass)
                            viewModel.increaseNumberOfSample(sampleClass)
                        } else {
                            if (viewModel.getCaptureMode() == false) {
                                classifyImage(image)
                            }
                        }
                        image.close()
                    }
                }

        // Must unbind the use-cases before rebinding them
        cameraProvider.unbindAll()

        try {
            // A variable number of use-cases can be passed here -
            // camera provides access to CameraControl & CameraInfo
            camera = cameraProvider.bindToLifecycle(
                this,
                cameraSelector,
                preview,
                imageAnalyzer
            )

            // Attach the viewfinder's surface provider to preview use case
            preview?.setSurfaceProvider(fragmentCameraBinding.viewFinder.surfaceProvider)
        } catch (exc: Exception) {
            Log.e(TAG, "Use case binding failed", exc)
        }
    }

    private fun classifyImage(image: ImageProxy) {
        // Copy out RGB bits to the shared bitmap buffer
        image.use { bitmapBuffer.copyPixelsFromBuffer(image.planes[0].buffer) }

        val imageRotation = image.imageInfo.rotationDegrees
        // Pass Bitmap and rotation to the transfer learning helper for
        // processing and classification.
        transferLearningHelper.classify(bitmapBuffer, imageRotation)
    }

    private fun addSample(image: ImageProxy, className: String) {
        // Copy out RGB bits to the shared bitmap buffer
        image.use { bitmapBuffer.copyPixelsFromBuffer(image.planes[0].buffer) }

        val imageRotation = image.imageInfo.rotationDegrees
        // Pass Bitmap and rotation to the transfer learning helper for
        // processing and prepare training data.
        transferLearningHelper.addSample(bitmapBuffer, className, imageRotation)
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onError(error: String) {
        activity?.runOnUiThread {
            Toast.makeText(requireContext(), error, Toast.LENGTH_SHORT).show()
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    override fun onResults(
        results: List<Category>?,
        inferenceTime: Long
    ) {
        activity?.runOnUiThread {
            // Update the result in inference mode.
            if (viewModel.getCaptureMode() == false) {
                // Show result
                results?.let { list ->
                    // Highlight the class which is highest score.
                    list.maxByOrNull { it.score }?.let {
                        highlightResult(it.label)
                    }
                    updateScoreClasses(list)
                }

                fragmentCameraBinding.tvInferenceTime.text =
                    String.format("%d ms", inferenceTime)
            }
        }
    }

    // Show the loss number after each training.
    override fun onLossResults(lossNumber: Float) {
        String.format(
            Locale.US,
            "Loss: %.3f", lossNumber
        ).let {
            fragmentCameraBinding.tvLossConsumerPause.text = it
            fragmentCameraBinding.tvLossConsumerResume.text = it
        }
    }

    // Show the accurate score of each class.
    private fun updateScoreClasses(categories: List<Category>) {
        categories.forEach {
            val view = getClassButtonScore(it.label)
            (view as? TextView)?.text = String.format(
                Locale.US, "%.1f", it.score
            )
        }
    }

    // Get the class button which represent for the label
    private fun getClassButton(label: String): View? {
        return when (label) {
            CLASS_ONE -> fragmentCameraBinding.llClassOne
            CLASS_TWO -> fragmentCameraBinding.llClassTwo
            CLASS_THREE -> fragmentCameraBinding.llClassThree
            CLASS_FOUR -> fragmentCameraBinding.llClassFour
            else -> null
        }
    }

    // Get the class button score which represent for the label
    private fun getClassButtonScore(label: String): View? {
        return when (label) {
            CLASS_ONE -> fragmentCameraBinding.tvNumberClassOne
            CLASS_TWO -> fragmentCameraBinding.tvNumberClassTwo
            CLASS_THREE -> fragmentCameraBinding.tvNumberClassThree
            CLASS_FOUR -> fragmentCameraBinding.tvNumberClassFour
            else -> null
        }
    }

    // Get the class name from resource id
    private fun getClassNameFromResourceId(id: Int): String {
        return when (id) {
            fragmentCameraBinding.llClassOne.id -> CLASS_ONE
            fragmentCameraBinding.llClassTwo.id -> CLASS_TWO
            fragmentCameraBinding.llClassThree.id -> CLASS_THREE
            fragmentCameraBinding.llClassFour.id -> CLASS_FOUR
            else -> {
                ""
            }
        }
    }

    // Highlight the current label and unhighlight the previous label
    private fun highlightResult(label: String?) {
        // skip the previous position if it is no position.
        previousClass?.let {
            setClassButtonHighlight(getClassButton(it), false)
        }
        if (label != null) {
            setClassButtonHighlight(getClassButton(label), true)
        }
        previousClass = label
    }

    private fun setClassButtonHighlight(view: View?, isHighlight: Boolean) {
        view?.run {
            background = AppCompatResources.getDrawable(
                context,
                if (isHighlight) R.drawable.btn_default_highlight else R.drawable.btn_default
            )
        }
    }

    // Update the number of samples. If there are no label in the samples,
    // set it 0.
    private fun updateNumberOfSample(numberOfSamples: Map<String, Int>) {
        fragmentCameraBinding.tvNumberClassOne.text = if (numberOfSamples
                .containsKey(CLASS_ONE)
        ) numberOfSamples.getValue(CLASS_ONE)
            .toString()
        else "0"
        fragmentCameraBinding.tvNumberClassTwo.text = if (numberOfSamples
                .containsKey(CLASS_TWO)
        ) numberOfSamples.getValue(CLASS_TWO)
            .toString()
        else "0"
        fragmentCameraBinding.tvNumberClassThree.text = if (numberOfSamples
                .containsKey(CLASS_THREE)
        ) numberOfSamples.getValue(CLASS_THREE)
            .toString()
        else "0"
        fragmentCameraBinding.tvNumberClassFour.text = if (numberOfSamples
                .containsKey(CLASS_FOUR)
        ) numberOfSamples.getValue(CLASS_FOUR)
            .toString()
        else "0"
    }

    private fun updateTrainingButtonState() {
        with(fragmentCameraBinding) {
            tvInferenceTime.visibility = if (viewModel
                    .getCaptureMode() == true
            ) View.GONE else View.VISIBLE

            btnCollectSample.visibility = if (
                viewModel.getTrainingState() == MainViewModel.TrainingState.PREPARE &&
                (viewModel.getNumberOfSample()?.size ?: 0) == 0 && viewModel
                    .getCaptureMode() == true
            ) View.VISIBLE else View.GONE

            btnStartTrain.visibility = if (
                viewModel.getTrainingState() == MainViewModel.TrainingState.PREPARE &&
                (viewModel.getNumberOfSample()?.size ?: 0) > 0 && viewModel
                    .getCaptureMode() == true
            ) View.VISIBLE else View.GONE

            btnPauseTrain.visibility =
                if (viewModel.getTrainingState() == MainViewModel
                        .TrainingState.TRAINING && viewModel
                        .getCaptureMode() == true
                ) View.VISIBLE else View.GONE

            btnResumeTrain.visibility =
                if (viewModel.getTrainingState() == MainViewModel
                        .TrainingState.PAUSE && viewModel
                        .getCaptureMode() == true
                ) View.VISIBLE else View.GONE

            // Disable adding button when it is training or in inference mode.
            (viewModel.getCaptureMode() == true && viewModel.getTrainingState() !=
                    MainViewModel.TrainingState.TRAINING).let { enable ->
                llClassOne.isClickable = enable
                llClassTwo.isClickable = enable
                llClassThree.isClickable = enable
                llClassFour.isClickable = enable
            }
        }
    }
}
